jobs/periodic/delete-unreachable-achievements/courses.ts

```diff
@@ -0,0 +1,78 @@
+import { getLogger } from '../../../common/logger/logger';
+import { EventRelationType, createRelation, parseRelation } from '../../../common/achievement/relation';
+import { prisma } from '../../../common/prisma';
+import tracer from '../../../common/logger/tracing';
+
+const logger = getLogger('CourseAchievementCleanup');
+
+const ParticipationGroupTemplate = 'pupil_course_participation';
+
+export const deleteUnreachableCourseAchievements = tracer.wrap('delete', _deleteUnreachableCourseAchievements);
+async function _deleteUnreachableCourseAchievements(dryRun: boolean = true) {
+    const till = new Date();
+    till.setDate(till.getDate() - 1);
+
+    const unachievedAchievementes = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: ParticipationGroupTemplate,
+            },
+            achievedAt: null,
+        },
+    });
+
+    const relationToUserIds: { [key: string]: string[] } = unachievedAchievementes.reduce((acc, curr) => {
+        if (!acc[curr.relation]) {
+            acc[curr.relation] = [];
+        }
+        acc[curr.relation].push(curr.userId);
+        return acc;
+    }, {});
+
+    logger.info('found possibly unachievable achievements', {
+        courses: relationToUserIds,
+        search_till: till,
+    });
+
+    for (const [relation, userIds] of Object.entries(relationToUserIds)) {
+        const [relationType, subcourseId] = parseRelation(relation);
+        if (relationType !== EventRelationType.Subcourse) {
+            logger.warn('unexpected relation type', {
+                relation,
+            });
+            continue;
+        }
+
+        const futureLectureCnt = await prisma.lecture.count({
+            where: {
+                subcourseId,
+                start: { gt: till },
+            },
+        });
+        if (futureLectureCnt > 0) {
+            logger.debug('skipping subcourse with future lectures', {
+                lecture_count: futureLectureCnt,
+                relation,
+                till,
+            });
+            continue;
+        }
+
+        logger.info('deleting pupil course participation achievements', {
+            relation,
+            user_ids: userIds,
+        });
+
+        if (!dryRun) {
+            await prisma.user_achievement.deleteMany({
+                where: {
```

Comment:
`achievedAt: null,` just to be on the safe side (and it could still be achieved while the job is running?)

Reply:
Actually, we even need to delete achieved achievements. When a user joins a course, the first achievement in the sequence will be marked as achieved, and the second will be created.  

This delete job will consider all courses that ended at most a day ago, ensuring that users can't achieve the second step in the sequence anymore.

Reply:
Ah, I see, maybe put as comment into the code given that this is non obvious? :)